### PR TITLE
Beacocleaning

### PR DESCRIPTION
**Pull request by Synpheros**

### CODE REVIEW

**Review by gorco — COMMENTED** [review]
Please, remove the package-lock.json from the repository

File: loginPlugins/beaconing.js [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+/**
+ * Default redirect value, if not specified.
+ *  When specified as a 'callback' query key this value is replaced.
+ *
+ *  This value is used to issue a callback to the initiator of the process with the
+ *  data usefull for the login (username, email, access control token).
+ * @type {string}
+ */
+var callback = 'https://localhost:3000/api/proxy/afront/loginbyplugin';
```
Comment (gorco) [inline loginPlugins/beaconing.js]: I think that callback, pluginId, pluginId and passport object fields should be load from a config file.

File: loginPlugins/beaconing.js [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    app.passport.serializeUser(function(user, done) {
+        done(null, user);
+    });
+
+    app.passport.deserializeUser(function(user, done) {
+        done(null, user);
+    });
+
+    var generateEmail = function(profile){
+        return profile.id + profile.username  + '@beaconing.eu';
```
Comment (gorco) [inline loginPlugins/beaconing.js]: What happen if the email already exist?

File: loginPlugins/beaconing.js [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+    var userExists = function (profile, db, callback) {
+
+        db.model('user').findOne({username: profile.username}, function (err, user) {
+            if (err) {
+                return callback(err);
+            }
+
+            if (!user) {
+                err = new Error('No account with that email address exists.');
```
Comment (gorco) [inline loginPlugins/beaconing.js]: You search a user by "username" but error say "no email address exists" and, also, should indicates what account doesn't exist.

File: loginPlugins/beaconing.js [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        done(null, user);
+    });
+
+    var generateEmail = function(profile){
+        return profile.id + profile.username  + '@beaconing.eu';
+    };
+
+    var userExists = function (profile, db, callback) {
+
+        db.model('user').findOne({username: profile.username}, function (err, user) {
```
Comment (gorco) [inline loginPlugins/beaconing.js]: Also, the username is unique, so you can use `find()` instead of `findOne()`.

File: loginPlugins/beaconing.js [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+        console.log(profile);
+
+        userExists(profile, app.db, function (err, user) {
+            if (err) {
+                // No user found, create a new user and assign a new role
+                console.info(err);
+                return addNewUser(profile, app.db, app.acl, done);
+            }
+
+            var userRoles = function(){
```
Comment (gorco) [inline loginPlugins/beaconing.js]: If the user exist you should return an error. Imagine that the user already exist, these user maybe is not the same that the new user that you are registering.

File: routes/users.js [hunk 6]
```diff
@@ -146,6 +146,67 @@ router.get(userIdRoute, authentication.authenticated, function (req, res, next)
     });
 });
 
+/**
+ * @api {get} /users/:domain/:externalId Gets the user information using externalId
```
Comment (gorco) [inline routes/users.js]: maybe the route should be `/users/domain/:domain/id/:externalId`

File: routes/users.js [hunk 7]
```diff
@@ -146,6 +146,67 @@ router.get(userIdRoute, authentication.authenticated, function (req, res, next)
...
+ *      }
+ *
+ * @apiError(400) UserNotFound No account with the given user id exists.
+ *
+ */
+router.get('/:domain/:externalId', authentication.authenticated, function (req, res, next) {
+    var domain = req.params.domain;
+    var externalId = req.params.externalId;
+
+    if(!domain){
```
Comment (gorco) [inline routes/users.js]: if domain or externalId not exist this get call doesn't happen. So you don't have to check it. You should check if domain and externalId are valid, but not if exist

File: routes/users.js [hunk 8]
```diff
@@ -147,7 +147,7 @@ router.get(userIdRoute, authentication.authenticated, function (req, res, next)
 });
 
 /**
- * @api {get} /users/:domain/:externalId Gets the user information using externalId
+ * @api {get} /users/external/:domain/:externalId Gets the user information using externalId
```
Comment (gorco) [inline routes/users.js:159]: This commit modifies code from the previous commit, please merge both. And I comment this line on previous commit. With "external" is better than before url

File: routes/users.js [hunk 9]
```diff
@@ -402,6 +402,61 @@ router.delete(userIdRoute, authentication.authenticated, function (req, res, nex
...
+ *      }
+ *
+ * @apiError(400) UserNotFound No account with the given external user id exists.
+ *
+ */
+router.delete('/external/:domain/:externalId', authentication.authenticated, function (req, res, next) {
+    var domain = req.params.domain;
+    var externalId = req.params.externalId;
+
+    if (!domain) {
```
Comment (gorco) [inline routes/users.js]: same that before (domain and expertalId exist always)

File: routes/applications.js [hunk 10]
```diff
@@ -471,22 +471,33 @@ router.put('/look/:prefix', authentication.authorized, function (req, res, next)
             return next(err);
         }
 
+        var users = [];
+        if (req.body.user) {
+            users.push(req.body.users);
+        }
+
+        if (req.body.users) {
```
Comment (gorco) [inline routes/applications.js:487]: is a new field? please, update doc

**Comment by gorco** [comment]
Done, please, check the commets. In general, nice pr

**Review by gorco — CHANGES_REQUESTED** [review]

File: loginPlugins/beaconing.js (hunk 1 above)
Comment (Synpheros) [reply loginPlugins/beaconing.js]: This is the default callback url. It is configurable as a query parameter in GET /login/beaconing?callback=$url

File: loginPlugins/beaconing.js (hunk 3 above)
Comment (Synpheros) [reply loginPlugins/beaconing.js]: This was corrected in a future commit

File: loginPlugins/beaconing.js (hunk 5 above)
Comment (Synpheros) [reply loginPlugins/beaconing.js]: I've made it configurable with a flag. By default it'll return an error!
Nice catch

File: loginPlugins/beaconing.js (hunk 4 above)
Comment (Synpheros) [reply loginPlugins/beaconing.js]: Stays like this as discussed in A16

File: routes/users.js (hunk 6 above)
Comment (Synpheros) [reply routes/users.js]: solved

File: routes/users.js (hunk 7 above)
Comment (Synpheros) [reply routes/users.js]: solved

File: routes/users.js (hunk 8 above)
Comment (Synpheros) [reply routes/users.js:159]: solved

File: routes/users.js (hunk 9 above)
Comment (Synpheros) [reply routes/users.js]: solved

**Comment by Synpheros** [comment]
Dear @gorco, all the changes have been added and tested!. I think it's ready to be merged 👍

**Comment by gorco** [comment]
Great, but I think there are two that have not been solved, please check the following comments:
"is a new field? please, update doc" (in /users.js)
"if domain or externalId not exist this get call doesn't happen. So you don't have to check it. You should check if domain and externalId are valid, but not if exist" (in /users.js)

**Pull request merged**